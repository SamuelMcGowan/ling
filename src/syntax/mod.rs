pub mod lexer;
pub mod parser;
pub mod token;

pub mod source;

use crate::chunk::Chunk;

#[derive(Debug)]
pub(crate) struct ParseContext<'a> {
    pub(crate) source: &'a str,
    pub(crate) chunk: &'a mut Chunk,
}

impl<'a> ParseContext<'a> {
    pub fn new(source: &'a str, chunk: &'a mut Chunk) -> Self {
        Self { source, chunk }
    }
}
