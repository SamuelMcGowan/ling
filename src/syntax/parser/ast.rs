use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
        serde::Deserialize,
    )];
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Expr,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),
    Unit,
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
        }
    }
}
