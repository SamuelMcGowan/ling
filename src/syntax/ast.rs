use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;
use crate::symbol_table::SymbolId;

use super::parser::ParseResult;
use super::source::Span;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
    )];
}

pub(crate) struct Spanned<T> {
    pub(crate) inner: T,
    pub(crate) span: Option<Span>,
}

impl<T> Spanned<ParseResult<T>> {
    pub fn transpose(self) -> ParseResult<Spanned<T>> {
        self.inner.map(|inner| Spanned {
            inner,
            span: self.span,
        })
    }
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
    Dummy,
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Block,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) struct Block {
    pub stmts: Vec<Stmt>,
    pub final_expr: Expr,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),

    Loop(Block),
    WhileLoop { cond: Expr, block: Block },

    Declaration { lhs: Ident, rhs: Expr },
    Assignment { lhs: Var, rhs: Expr },

    Dummy,
}

impl Stmt {
    pub fn expect_delim(&self) -> bool {
        match self {
            Self::Expr(expr) => expr.expect_delim(),

            Self::Loop(_) | Self::WhileLoop { .. } => false,

            Self::Declaration { .. } => true,
            Self::Assignment { .. } => true,

            // the recovery function should have recovered past any semicolons
            Self::Dummy => false,
        }
    }
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),

    UnaryOp {
        op: UnaryOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Call {
        callee: Box<Expr>,
        args: Vec<Expr>,
    },

    If {
        branches: Vec<IfBranch>,
        else_: Option<Box<Block>>,
    },

    Var(Var),

    Unit,

    Dummy,
}

impl Expr {
    pub fn expect_delim(&self) -> bool {
        !matches!(self, Self::If { .. })
    }
}

#[derive(Node!, Copy)]
pub enum UnaryOp {
    Neg,
    Not,
}

#[derive(Node!, Copy)]
pub enum BinOp {
    LogicalAnd,
    LogicalOr,

    Equal,
    NotEqual,
    Gt,
    Lt,
    GtEqual,
    LtEqual,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Pow,

    Call,
    Access,
}

#[derive(Node!)]
pub(crate) struct IfBranch {
    pub(crate) cond: Expr,
    pub(crate) then: Block,
}

#[derive(Node!)]
pub(crate) enum Var {
    Simple(Ident),
    Field { expr: Box<Expr>, field: Ident },
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
    Resolved(SymbolId),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
            Self::Resolved(_) => None,
        }
    }
}
