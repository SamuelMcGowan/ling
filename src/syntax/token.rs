use super::span::Span;
use crate::chunk::ConstIdx;

#[derive(Debug, Clone, Copy)]
pub(crate) struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum TokenKind {
    Punct(Punct),
    Keyword(Keyword),

    Ident(Ustr),
    Const(ConstIdx),

    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Punct {
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,
    Match,

    If,
    Elif,
    Else,

    Loop,
    While,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Error {
    UnexpectedChar(char),

    InvalidEscape,
    UnterminatedString,

    IntegerOverflow,
    MissingDigits,
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::syntax::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (kwd $kwd:ident) => {{
        use crate::syntax::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};

    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::syntax::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (constant $constant:literal) => {{
        use crate::syntax::token::*;
        use crate::chunk::ConstIdx;

        TokenKind::Const(ConstIdx($constant))
    }};

    (error $error:ident $($value:tt)?) => {{
        use crate::syntax::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
