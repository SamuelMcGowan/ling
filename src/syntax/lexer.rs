use ustr::Ustr;

use super::span::Span;
use super::token::{tkind, Token, TokenKind};
use super::{Cursor, ParseContext};
use crate::value::Value;

impl<'a> ParseContext<'a> {
    pub fn lex_tokens(&mut self) -> Lexer<'_, 'a> {
        Lexer {
            cursor: Cursor::new(self.source),
            context: self,
        }
    }
}

pub(crate) struct Lexer<'ctx, 'a> {
    cursor: Cursor<'a>,
    context: &'ctx mut ParseContext<'a>,
}

impl Lexer<'_, '_> {
    fn lex_token(&mut self) -> Option<Token> {
        loop {
            let start_pos = self.cursor.byte_pos();
            let c = self.cursor.next()?;

            let kind = match c {
                c if c.is_ascii_whitespace() => continue,
                '#' => {
                    while !matches!(self.cursor.next(), Some('\n') | None) {}
                    continue;
                }

                '(' => tkind!(punct LParen),
                ')' => tkind!(punct RParen),
                '{' => tkind!(punct LBrace),
                '}' => tkind!(punct RBrace),
                '[' => tkind!(punct LBracket),
                ']' => tkind!(punct RBracket),

                '.' => tkind!(punct Dot),
                ',' => tkind!(punct Comma),
                ':' => tkind!(punct Colon),
                ';' => tkind!(punct Semicolon),

                '-' if self.cursor.eat('>') => tkind!(punct Arrow),
                '=' if self.cursor.eat('>') => tkind!(punct FatArrow),

                '+' if self.cursor.eat('=') => tkind!(punct AddEqual),
                '-' if self.cursor.eat('=') => tkind!(punct SubEqual),
                '*' if self.cursor.eat('=') => tkind!(punct MulEqual),
                '/' if self.cursor.eat('=') => tkind!(punct DivEqual),
                '%' if self.cursor.eat('=') => tkind!(punct ModEqual),
                '^' if self.cursor.eat('=') => tkind!(punct PowEqual),

                '+' => tkind!(punct Add),
                '-' => tkind!(punct Sub),
                '*' => tkind!(punct Mul),
                '/' => tkind!(punct Div),
                '%' => tkind!(punct Mod),
                '^' => tkind!(punct Pow),

                '>' if self.cursor.eat('=') => tkind!(punct GtEqual),
                '<' if self.cursor.eat('=') => tkind!(punct LtEqual),
                '=' if self.cursor.eat('=') => tkind!(punct EqualEqual),

                '>' => tkind!(punct Gt),
                '<' => tkind!(punct Lt),
                '=' => tkind!(punct Equal),

                _ if is_ident_start(c) => self.lex_ident(start_pos),
                '0'..='9' => self.lex_number(c as u8),
                '"' => self.lex_string(),

                _ => tkind!(error UnexpectedChar(c)),
            };

            let span = Span::new(start_pos, self.cursor.byte_pos());

            return Some(Token { span, kind });
        }
    }

    fn lex_ident(&mut self, start_pos: usize) -> TokenKind {
        while matches!(self.cursor.peek(), Some(c) if is_ident(c)) {
            self.cursor.next();
        }

        let lexeme = &self.cursor.as_str_all()[start_pos..self.cursor.byte_pos()];

        match lexeme {
            "func" => tkind!(kwd Func),
            "return" => tkind!(kwd Return),

            "type" => tkind!(kwd Type),
            "enum" => tkind!(kwd Enum),
            "data" => tkind!(kwd Data),
            "typeclass" => tkind!(kwd Typeclass),
            "impl" => tkind!(kwd Impl),

            "let" => tkind!(kwd Let),
            "match" => tkind!(kwd Match),

            "if" => tkind!(kwd If),
            "elif" => tkind!(kwd Elif),
            "else" => tkind!(kwd Else),

            "loop" => tkind!(kwd Loop),
            "while" => tkind!(kwd While),

            "true" => self.add_constant(Value::Bool(true)),
            "false" => self.add_constant(Value::Bool(false)),

            _ => tkind!(ident lexeme),
        }
    }

    fn lex_string(&mut self) -> TokenKind {
        let mut s = String::new();

        let mut unterminated_string = true;
        let mut invalid_escape = false;

        while let Some(c) = self.cursor.next() {
            let c = match c {
                '"' => {
                    unterminated_string = false;
                    break;
                }
                '\\' => match self.cursor.next() {
                    Some(c) => match c {
                        'n' => '\n',
                        'r' => '\n',
                        't' => '\t',
                        '0' => '\0',
                        _ => {
                            invalid_escape = true;
                            c
                        }
                    },
                    None => return tkind!(error UnterminatedString),
                },
                _ => c,
            };
            s.push(c);
        }

        if unterminated_string {
            return tkind!(error UnterminatedString);
        }
        if invalid_escape {
            return tkind!(error InvalidEscape);
        }

        let interned = Ustr::from(&s);
        self.add_constant(Value::String(interned))
    }

    fn lex_number(&mut self, first: u8) -> TokenKind {
        let int_digits = self.lex_digits(Some(first));

        if self.cursor.eat('.') {
            let fraction_digits = self.lex_digits(None);

            let mut exponent_overflow = false;
            let mut exponent_missing_digits = false;

            let exponent: i32 = if self.cursor.eat('e') || self.cursor.eat('E') {
                let sign = if self.cursor.eat('-') {
                    -1
                } else {
                    self.cursor.eat('+');
                    1
                };

                let exponent_digits = self.lex_digits(None);

                let exponent = parse_int(&exponent_digits).and_then(|n| n.try_into().ok());
                let exponent: i32 = match exponent {
                    Some(exponent) => exponent,
                    None => {
                        exponent_overflow = true;
                        0
                    }
                };

                exponent_missing_digits = exponent_digits.is_empty();

                exponent * sign
            } else {
                0
            };

            if exponent_overflow {
                return tkind!(error IntegerOverflow);
            }

            // integer won't have missing digits
            if fraction_digits.is_empty() || exponent_missing_digits {
                return tkind!(error MissingDigits);
            }

            let int_digits = trim_leading_zeros(&int_digits).iter();
            let fraction_digits = trim_trailing_zeros(&fraction_digits).iter();

            let f: f64 = minimal_lexical::parse_float(int_digits, fraction_digits, exponent);
            self.add_constant(Value::Float(f))
        } else {
            let Some(n) = parse_int(&int_digits) else {
                return tkind!(error IntegerOverflow);
            };
            self.add_constant(Value::Integer(n))
        }
    }

    fn lex_digits(&mut self, first: Option<u8>) -> Vec<u8> {
        let mut digits = vec![];
        digits.extend(first);
        loop {
            match self.cursor.peek() {
                Some(c @ '0'..='9') => {
                    digits.push(c as u8);
                    self.cursor.next();
                }
                Some('_') => continue,
                _ => break,
            }
        }
        digits
    }

    fn add_constant(&mut self, value: Value) -> TokenKind {
        let idx = self.context.chunk.add_constant(value);
        TokenKind::Const(idx)
    }
}

impl Iterator for Lexer<'_, '_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        self.lex_token()
    }
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_ident(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn parse_int(digits: &[u8]) -> Option<u64> {
    let mut n = 0u64;
    for &digit in digits {
        let digit = (digit - b'0') as u64;
        n = n.checked_mul(10)?;
        n = n.checked_add(digit)?;
    }
    Some(n)
}

fn trim_leading_zeros(mut digits: &[u8]) -> &[u8] {
    while let [b'0', rest @ ..] = digits {
        digits = rest;
    }
    digits
}

fn trim_trailing_zeros(mut digits: &[u8]) -> &[u8] {
    while let [rest @ .., b'0'] = digits {
        digits = rest;
    }
    digits
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::chunk::{Chunk, ConstIdx};

    #[test]
    fn punct_single() {
        check_tokens(
            "+-:;",
            &[
                tkind!(punct Add),
                tkind!(punct Sub),
                tkind!(punct Colon),
                tkind!(punct Semicolon),
            ],
        );
    }

    #[test]
    fn punct_double() {
        check_tokens(
            "->==+=",
            &[
                tkind!(punct Arrow),
                tkind!(punct EqualEqual),
                tkind!(punct AddEqual),
            ],
        );
    }

    #[test]
    fn whitespace() {
        check_tokens(" .\r\n. ", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn comment() {
        check_tokens(
            ".# comment\r\n.# comment",
            &[tkind!(punct Dot), tkind!(punct Dot)],
        );
    }

    #[test]
    fn comment_empty() {
        check_tokens(".#\r\n.#", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn keyword() {
        check_tokens("else", &[tkind!(kwd Else)]);
    }

    #[test]
    fn ident() {
        check_tokens("else_", &[tkind!(ident "else_")])
    }

    #[test]
    fn literals() {
        check_constant("true", Value::Bool(true));
        check_constant("false", Value::Bool(false));
    }

    #[test]
    fn string() {
        check_constant("\"hello\"", Value::String(Ustr::from("hello")));
        check_constant("\"\"", Value::String(Ustr::from("")));
    }

    #[test]
    fn string_unterminated() {
        check_tokens("\"hello", &[tkind!(error UnterminatedString)]);
        check_tokens("\"", &[tkind!(error UnterminatedString)]);
    }

    #[test]
    fn integer() {
        check_constant("123", Value::Integer(123));
    }

    #[test]
    fn float() {
        check_constant("12.34e-5", Value::Float(12.34e-5));
    }

    #[test]
    fn missing_digits() {
        check_tokens("12.", &[tkind!(error MissingDigits)]);
        check_tokens("12.1e", &[tkind!(error MissingDigits)]);
    }

    fn check_tokens(s: &str, t: &[TokenKind]) {
        let mut chunk = Chunk::default();
        let mut context = ParseContext::new(s, &mut chunk);
        let tokens: Vec<_> = context.lex_tokens().map(|token| token.kind).collect();
        assert_eq!(&tokens, t)
    }

    fn check_constant(s: &str, value: Value) {
        let mut chunk = Chunk::default();
        let mut context = ParseContext::new(s, &mut chunk);

        let tokens: Vec<_> = context.lex_tokens().map(|token| token.kind).collect();
        assert_eq!(&tokens, &[tkind!(constant 0)]);

        let value_found = chunk.get_constant(ConstIdx(0)).expect("constant not found");
        assert!(value_found.is_equal(&value));
    }
}
