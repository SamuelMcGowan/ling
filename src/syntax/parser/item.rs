use crate::syntax::ast::*;
use crate::syntax::token::tkind;

use super::*;

impl Parser<'_> {
    pub fn parse_module(&mut self) -> ParseResult<Module> {
        let mut items = vec![];

        while !self.tokens.at_end() {
            let item = self.parse_item_or_recover();
            items.push(item);
        }

        Ok(Module { items })
    }

    fn parse_item_or_recover(&mut self) -> Item {
        self.parse_or_recover(Self::parse_item, |parser| {
            parser.recover_until(&[tkind!(kwd Func)]);
            Item::Dummy
        })
    }

    fn parse_item(&mut self) -> ParseResult<Item> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) if token.kind == tkind!(kwd Func) => {
                Ok(Item::Func(self.parse_func()?))
            }
            other => Err(ParseError::unexpected("an item", other)),
        }
    }

    fn parse_func(&mut self) -> ParseResult<Func> {
        let ident = self.parse_ident()?;

        let params = {
            let tokens = self.expect_group(BracketKind::Round, "parameters list")?;
            let mut parser = self.parser_for_tokens(tokens);

            parser.parse_list("function arguments", tkind!(punct Comma), |parser| {
                let ident = parser.parse_ident()?;
                parser.expect_kind(tkind!(punct Colon))?;
                let ty = parser.parse_ty()?;
                Ok((ident, ty))
            })?
        };

        let ret_ty = if self.eat_kind(tkind!(punct Arrow)) {
            self.parse_ty()?
        } else {
            Ty::Unit
        };

        let body = self.parse_block()?;

        Ok(Func {
            ident,
            params,
            ret_ty,
            body,
        })
    }

    fn parse_ident(&mut self) -> ParseResult<Ident> {
        match self.tokens.next() {
            Some(TokenTree::Token(Token {
                kind: TokenKind::Ident(ident),
                ..
            })) => Ok(Ident::Unresolved(ident)),
            other => Err(ParseError::unexpected("an identifier", other)),
        }
    }

    fn parse_ty(&mut self) -> ParseResult<Ty> {
        match self.tokens.next() {
            Some(TokenTree::Token(Token {
                kind: TokenKind::Ident(ident),
                ..
            })) => {
                let params = if let Some(tokens) = self.eat_group(BracketKind::Square) {
                    let mut parser = self.parser_for_tokens(tokens);
                    parser.parse_list("type parameters", tkind!(punct Comma), Parser::parse_ty)?
                } else {
                    vec![]
                };

                Ok(Ty::Constructed {
                    ident: Ident::Unresolved(ident),
                    params,
                })
            }

            Some(TokenTree::Group {
                bracket_kind: BracketKind::Round,
                tokens,
                ..
            }) if tokens.is_empty() => Ok(Ty::Unit),

            other => Err(ParseError::unexpected("a type", other)),
        }
    }

    pub(super) fn parse_block(&mut self) -> ParseResult<Block> {
        let tokens = self.expect_group(BracketKind::Curly, "block")?;
        let mut parser = self.parser_for_tokens(tokens);

        let mut stmts = vec![];

        let mut eval_as_final = false;

        let mut final_stmt_span = None;

        while !parser.tokens.at_end() {
            let stmt = parser.parse_spanned(Parser::parse_stmt_or_recover);
            final_stmt_span = stmt.span;

            let expect_delim = stmt.inner.expect_delim();
            let found_delim = parser.eat_kind(tkind!(punct Semicolon));

            stmts.push(stmt.inner);

            if expect_delim && !found_delim {
                // peek so that we can carry on parsing statements
                if let Some(tt) = parser.tokens.peek() {
                    parser.report(ParseError::unexpected("semicolon", Some(tt.clone())));
                }
                // don't break, we want to keep parsing
            }

            eval_as_final = !found_delim;
        }

        // Guards against accidentally using `self` instead of `parser`.
        drop(parser);

        let final_expr = if eval_as_final {
            match stmts.pop() {
                Some(Stmt::Expr(expr)) => expr,
                Some(stmt) => {
                    // this is ok to unwrap since statement will always consume
                    // at least one token
                    self.report(ParseError::InvalidImplicitReturn(final_stmt_span.unwrap()));

                    // add the statement back to the block to improve analysis
                    stmts.push(stmt);

                    Expr::Dummy
                }
                None => Expr::Unit,
            }
        } else {
            Expr::Unit
        };

        Ok(Block { stmts, final_expr })
    }

    fn parse_stmt_or_recover(&mut self) -> Stmt {
        self.parse_or_recover(Self::parse_stmt, |parser| {
            // Recover *past* the semicolon so that we don't have to
            // guess whether or not to expect a semicolon in `parse_block`.

            // We don't recover up to any keywords in case they appear in the statement.
            // This could change if we later support top-items inside blocks, since these can't
            // appear inside expressions (except inside groups).

            parser.recover_past(tkind!(punct Semicolon));

            Stmt::Dummy
        })
    }

    fn parse_stmt(&mut self) -> ParseResult<Stmt> {
        let stmt_spanned = self
            .parse_spanned(|p| match p.tokens.peek() {
                Some(TokenTree::Token(t)) if t.kind == tkind!(kwd Loop) => {
                    p.tokens.next();
                    p.parse_block().map(Stmt::Loop)
                }
                Some(TokenTree::Token(t)) if t.kind == tkind!(kwd While) => {
                    p.tokens.next();
                    let cond = p.parse_expr()?;
                    let block = p.parse_block()?;
                    Ok(Stmt::WhileLoop { cond, block })
                }
                Some(TokenTree::Token(t)) if t.kind == tkind!(kwd Let) => {
                    p.tokens.next();
                    let ident = p.parse_ident()?;
                    p.expect_kind(tkind!(punct Equal))?;
                    let rhs = p.parse_expr()?;
                    Ok(Stmt::Declaration { lhs: ident, rhs })
                }
                _ => p.parse_expr().map(Stmt::Expr),
            })
            .transpose()?;

        if self.eat_kind(tkind!(punct Equal)) {
            let rhs = self.parse_expr()?;
            match stmt_spanned.inner {
                Stmt::Expr(Expr::Var(var)) => Ok(Stmt::Assignment { lhs: var, rhs }),
                // parse_expr will always consume at least one token so
                // it's ok to unwrap the span
                _ => Err(ParseError::InvalidAssignmentTarget(
                    stmt_spanned.span.unwrap(),
                )),
            }
        } else {
            Ok(stmt_spanned.inner)
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use crate::syntax::parser::test_parse;

    #[test]
    fn simple_func() {
        assert_ron_snapshot!(test_parse("foo() {}", |p| p.parse_func()));
    }

    #[test]
    fn func_with_args() {
        assert_ron_snapshot!(test_parse("foo(a: uint, b: uint) -> string {}", |p| p.parse_func()));
    }

    #[test]
    fn func_with_args_and_trailing_comma() {
        assert_ron_snapshot!(test_parse("foo(a: uint, b: uint,) {}", |p| p.parse_func()));
    }

    #[test]
    fn func_with_missing_rparen() {
        assert_ron_snapshot!(test_parse("foo(a: uint, {}", |p| p.parse_func()));
    }

    #[test]
    fn func_with_missing_ret_type() {
        assert_ron_snapshot!(test_parse("foo(a: uint) -> {}", |p| p.parse_func()));
    }

    #[test]
    fn func_recovery() {
        assert_ron_snapshot!(test_parse(
            "func foo(a: uint) -> {} {} func my_func() {}",
            |p| p.parse_module()
        ));
    }

    #[test]
    fn block_empty() {
        assert_ron_snapshot!(test_parse("{}", |p| p.parse_block()));
    }

    #[test]
    fn block_final_expr() {
        assert_ron_snapshot!(test_parse("{12}", |p| p.parse_block()));
    }

    #[test]
    fn block_no_final_expr() {
        assert_ron_snapshot!(test_parse("{12;}", |p| p.parse_block()));
    }

    #[test]
    fn block_two_exprs_final_expr() {
        assert_ron_snapshot!(test_parse("{14;12}", |p| p.parse_block()));
    }

    #[test]
    fn block_two_exprs_no_final_expr() {
        assert_ron_snapshot!(test_parse("{14;12;}", |p| p.parse_block()));
    }

    #[test]
    fn block_errors() {
        assert_ron_snapshot!(test_parse("{10; 11 12 13;}", |p| p.parse_block()));
    }

    #[test]
    fn block_errors_final_expr() {
        assert_ron_snapshot!(test_parse("{10; 11 12}", |p| p.parse_block()));
    }

    #[test]
    fn block_undelimited() {
        assert_ron_snapshot!(test_parse("{ if a { 12 } if b { 12 } }", |p| p.parse_block()));
    }

    #[test]
    fn assignment() {
        assert_ron_snapshot!(test_parse("a = 12", |p| p.parse_stmt()));
    }

    #[test]
    fn field_assignment() {
        assert_ron_snapshot!(test_parse("my_struct.field = 12", |p| p.parse_stmt()));
    }

    #[test]
    fn invalid_assignment_target() {
        assert_ron_snapshot!(test_parse("a + b = 12", |p| p.parse_stmt()));
    }

    #[test]
    fn loop_assignment_target() {
        assert_ron_snapshot!(test_parse("loop {} = 12", |p| p.parse_stmt()));
    }

    #[test]
    fn return_assignment() {
        assert_ron_snapshot!(test_parse("{ a = b }", |p| p.parse_block()));
    }

    #[test]
    fn let_stmt() {
        assert_ron_snapshot!(test_parse("let a = 12 + 3", |p| p.parse_stmt()));
    }
}
