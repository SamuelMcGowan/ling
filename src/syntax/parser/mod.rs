pub mod expr;
pub mod item;

use super::ast::Spanned;
use super::source::Span;
use super::token::{BracketKind, Token, TokenKind};
use super::token_stream::{TokenIter, TokenTree};

#[derive(Debug, Clone, serde::Serialize)]
pub(crate) enum ParseError {
    Unexpected {
        expected: String,
        found: Option<Span>,
    },
    InvalidAssignmentTarget(Span),
    InvalidImplicitReturn(Span),
}

impl ParseError {
    fn unexpected(expected: impl Into<String>, found: Option<TokenTree>) -> Self {
        Self::Unexpected {
            expected: expected.into(),
            found: found.map(|tt| tt.span()),
        }
    }
}

pub(crate) type ParseResult<T> = Result<T, ParseError>;

pub(crate) struct Parser<'a> {
    errors: &'a mut Vec<ParseError>,
    tokens: TokenIter,
}

// Utilities.
impl<'a> Parser<'a> {
    pub fn new(tokens: TokenIter, errors: &'a mut Vec<ParseError>) -> Self {
        Self { tokens, errors }
    }

    fn parser_for_tokens(&mut self, tokens: TokenIter) -> Parser {
        Parser {
            errors: self.errors,
            tokens,
        }
    }

    fn eat_kind(&mut self, kind: TokenKind) -> bool {
        match self.tokens.peek() {
            Some(TokenTree::Token(token)) if token.kind == kind => {
                self.tokens.next();
                true
            }
            _ => false,
        }
    }

    fn expect_kind(&mut self, kind: TokenKind) -> ParseResult<Token> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) if token.kind == kind => Ok(token),
            other => Err(ParseError::unexpected(
                format!("token of kind {kind:?}"),
                other,
            )),
        }
    }

    fn expect_group(
        &mut self,
        kind: BracketKind,
        expected: impl Into<String>,
    ) -> ParseResult<TokenIter> {
        match self.tokens.next() {
            Some(TokenTree::Group {
                bracket_kind,
                tokens,
                ..
            }) if bracket_kind == kind => Ok(tokens.into_iter()),
            other => Err(ParseError::unexpected(expected.into(), other)),
        }
    }

    fn expect_end(&mut self, tree_name: &str) -> ParseResult<()> {
        match self.tokens.next() {
            None => Ok(()),
            Some(tree) => Err(ParseError::unexpected(
                format!("end of {tree_name}"),
                Some(tree),
            )),
        }
    }

    fn eat_group(&mut self, kind: BracketKind) -> Option<TokenIter> {
        match self.tokens.peek() {
            Some(TokenTree::Group { bracket_kind, .. }) if *bracket_kind == kind => {
                let Some(TokenTree::Group { tokens, ..}) = self.tokens.next() else {
                    unreachable!();
                };

                Some(tokens.into_iter())
            }
            _ => None,
        }
    }

    fn parse_list<T>(
        &mut self,
        tree_name: &str,
        delim: TokenKind,
        parse: impl Fn(&mut Self) -> ParseResult<T>,
    ) -> ParseResult<Vec<T>> {
        let mut items = vec![];

        while !self.tokens.at_end() {
            let item = parse(self)?;
            items.push(item);

            if !self.eat_kind(delim) {
                if let Some(token) = self.tokens.next() {
                    return Err(ParseError::unexpected(
                        format!("token of kind {delim:?} or end of {tree_name:?}"),
                        Some(token),
                    ));
                }
                break;
            }
        }

        Ok(items)
    }
}

// Error handling/recovery.
impl Parser<'_> {
    fn parse_or_recover<T>(
        &mut self,
        parser: impl Fn(&mut Self) -> ParseResult<T>,
        recover: impl Fn(&mut Self) -> T,
    ) -> T {
        parser(self).unwrap_or_else(|error| {
            self.report(error);
            recover(self)
        })
    }

    fn parse_spanned<T>(&mut self, parser: impl Fn(&mut Self) -> T) -> Spanned<T> {
        let start = self.tokens.peek().map(|tt| tt.span());
        let inner = parser(self);
        let end = self.tokens.prev_span();

        let span = match (start, end) {
            (Some(start), Some(end)) => Some(start.union(end)),
            _ => None,
        };

        Spanned { inner, span }
    }

    fn recover_until(&mut self, kinds: &[TokenKind]) {
        while let Some(tt) = self.tokens.peek() {
            if matches!(tt, TokenTree::Token(token) if kinds.contains(&token.kind)) {
                break;
            }
            self.tokens.next();
        }
    }

    fn recover_past(&mut self, kind: TokenKind) {
        self.tokens
            .by_ref()
            .take_while(|token| !matches!(token, TokenTree::Token(token) if token.kind == kind))
            .for_each(|_| {});
    }

    fn report(&mut self, error: ParseError) {
        self.errors.push(error);
    }
}

#[cfg(test)]
pub(crate) fn test_parse<T>(
    source: &str,
    f: impl Fn(&mut Parser) -> T,
) -> (T, Vec<Token>, Vec<ParseError>) {
    use crate::syntax::lexer::Lexer;
    use crate::syntax::token_stream::TokenStream;

    let lexer = Lexer::new(source);
    let (tokens, mismatched_brackets) = TokenStream::from_lexer(lexer);

    let mut errors = vec![];
    let mut parser = Parser::new(tokens.into_iter(), &mut errors);
    let res = f(&mut parser);

    (res, mismatched_brackets, errors)
}
