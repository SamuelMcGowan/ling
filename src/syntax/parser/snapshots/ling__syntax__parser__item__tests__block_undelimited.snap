---
source: src/syntax/parser/item.rs
expression: "test_parse(\"{ if a { 12 } if b { 12 } }\", |p| p.parse_block())"
---
(Ok(Block(
  stmts: [
    Expr(If(
      branches: [
        IfBranch(
          cond: Var(Unresolved("a")),
          then: Block(
            stmts: [
              Expr(Const(ConstIdx(0))),
            ],
            eval_as_final: true,
          ),
        ),
      ],
      else_: None,
    )),
    Expr(If(
      branches: [
        IfBranch(
          cond: Var(Unresolved("b")),
          then: Block(
            stmts: [
              Expr(Const(ConstIdx(1))),
            ],
            eval_as_final: true,
          ),
        ),
      ],
      else_: None,
    )),
  ],
  eval_as_final: true,
)), [], [])
